contributors/design-proposals/scheduling/scheduling-policy.md

```diff
@@ -0,0 +1,355 @@
+
+### Scheduler name
+
+It should be possible to allow users to use only specific schedulers using `allowedSchedulerNames` field.
+
+If `allowedSchedulerNames` is absent from SchedulingPolicy, no scheduler is allowed by this specific policy.
+
+#### Examples
+
+Allow serviceaccounts to use either the default-scheduler (which is used by specifying `spec.schedulerName` in pod definition) or the `my-scheduler` scheduler (by specifying `spec.schedulerName: "my-scheduler"`):
+```yaml
+Kind: SchedulingPolicy
+spec:
+  allowedSchedulerNames:
+  - default-scheduler
+  - my-scheduler
+```
+
+
+Allow all schedulers:
+```yaml
+Kind: SchedulingPolicy
+spec:
+  allowedSchedulerNames: []
+```
+
+
+### Tolerations
+
+Toleration usage can be allowed using fine-grain rules with `allowedTolerations` field. If specifying multiple `allowedTolerations`, pod will be scheduled if one of the allowedTolerations is satisfied.
+
+If `allowedTolerations` is absent from SchedulingPolicy, no toleration is allowed.
+
+#### Examples
+
+##### Fine-grain allowedTolerations
+```yaml
+Kind: SchedulingPolicy
+spec:
+  allowedTolerations:
+  - keys: ["mykey"]
+    operators: ["Equal"]
+    values: ["value"]
+    effects: ["NoSchedule"]
+  - keys: ["other_key"]
+    operators: ["Exists"]
+    effects: ["NoExecute"]
+```
+This example allows tolerations in the following forms:
+- tolerations that tolerates taints with key named `mykey` that has a value `value` and with a `NoSchedule` effect.
+- tolerations that tolerates taints with key `other_key` that has a `NoExecute` effect.
+
+##### Coarse-grain allowedTolerations
+```yaml
+Kind: SchedulingPolicy
+spec:
+  allowedTolerations:
+  - keys: []
+    operators: []
+    values: []
+    effects: ["PreferNoSchedule"]
+  - keys: []
+    operators: ["Exists"]
+    effects: ["NoSchedule"]
+```
+This example allows tolerations in the following forms:
+- tolerations that tolerates all `PreferNoSchedule` taints with any value.
+- tolerations that tolerates taints based on any key existence with effect `NoSchedule`.
+Also note that this SchedulingPolicy does not allow tolerating NoExecute taints.
+
+
+### Priority classes
+
+We must be able to enforce users to use specific priority classes by using the `allowedPriorityClasseNames` field.
+
+If `allowedPriorityClasseNames` is absent from SchedulingPolicy, no priority class is allowed.
+
+#### Examples
+
+##### Only allow a single priority class
+```yaml
+Kind: SchedulingPolicy
+spec:
+  allowedPriorityClasseNames:
+  - high-priority
+```
+In this example, only the `high-priority` PriorityClass is allowed.
+
+
+##### Allow all priorities
+
+```yaml
+Kind: SchedulingPolicy
+spec:
+  allowedPriorityClasseNames: []
+```
+In this example, all priority classes are allowed.
+
+### Node Selector
+
+As anti-affinity rules are really time-consuming, we must be able to restrict their usage with `allowedNodeSelectors`.
+
+If `allowedNodeSelectors` is totally absent from the spec, no node selector is allowed.
```

Comment:
This doesn't make sense. A pod with no nodeSelector targets the most nodes possible. Adding more selectors constrains a pod. 

Generally, you want to require a set of nodeSelector labels be present, error if the pod tries to specify nodeSelector components that conflict with that required set, and allow the pod to specify any additional nodeSelector components it wants. That is what the current podnodeselector admission plugin does

Reply:
Do you think we could do it this way?
I'm taking `nodeSelector` as a simple example to start with:
```yaml
apiVersion: extensions/valpha1
kind: SchedulingPolicy
metadata:
  name: my-schedpol
spec:
  nodeSelectors:
    required:
      beta.kubernetes.io/arch: ["amd64", "arm"] # pick one of thoses mandadory values
    default:
      beta.kubernetes.io/os: amd64 # Here is the default value unless specified
    allowed:
      failure-domain.beta.kubernetes.io/region: [] # any value can be sepcified
```

Given the deny-by-default design, some kind of `forbiden` subsection actually would'nt make sense.

Reply:
@arnaudmz - I agree, given the design, `forbiden` doesn’t make much sense.